backend/typescript/services/implementations/ReviewService.ts

```diff
@@ -0,0 +1,174 @@
+              {
+                review_id: newReview.id,
+                title_prefix: book.titlePrefix,
+                title: book.title,
+                series_id: series?.id || null,
+                series_order: book.seriesOrder || null,
+                illustrator: book.illustrator || null,
+                translator: book.translator || null,
+                formats: book.formats,
+                age_range: [book.minAge, book.maxAge],
+              },
+              { transaction: t },
+            );
+
+            const authorsRet: Author[] = await Promise.all(
+              book.authors.map(async (a) => {
+                const author = await PgAuthor.findOrCreate({
+                  where: {
+                    full_name: a.fullName,
+                    display_name: a.displayName || null,
+                    attribution: a.attribution || null,
+                  },
+                  transaction: t,
+                }).then((data) => data[0]);
+                newBook.$add("authors", author, { transaction: t });
+
+                return {
+                  fullName: author.full_name,
+                  displayName: author.display_name,
+                  attribution: author.attribution,
+                };
+              }),
+            );
+
+            const publishersRet: Publisher[] = await Promise.all(
+              book.publishers.map(async (p) => {
+                publisher = await PgPublisher.findOrCreate({
+                  where: {
+                    full_name: p.fullName,
+                    publish_year: p.publishYear,
+                  },
+                  transaction: t,
+                }).then((data) => data[0]);
+                return {
+                  fullName: publisher.full_name,
+                  publishYear: publisher.publish_year,
+                };
+              }),
+            );
+
+            await newBook.$add("publishers", publisher, { transaction: t });
+            await newReview.$add("books", newBook, { transaction: t });
+
+            return {
+              title: newBook.title,
+              titlePrefix: newBook.title_prefix,
+              seriesOrder: newBook.series_order,
+              illustrator: newBook.illustrator,
+              translator: newBook.translator,
+              formats: newBook.formats,
+              minAge: newBook.age_range[0].value,
+              maxAge: newBook.age_range[1].value,
+              authors: authorsRet,
+              publishers: publishersRet,
+              seriesName: series?.name || null,
+            };
+          }),
+        );
+
+        return {
+          reviewId: newReview.id,
+          body: newReview.body,
+          coverImages: newReview.cover_images,
+          byline: newReview.byline,
+          featured: newReview.featured,
+          books: booksRet,
+          tags: tagsRet,
+          updatedAt: newReview.updatedAt.getTime(),
+          publishedAt: newReview.published_at.getTime() / 1000,
```

Comment:
Will this fail if published_at is null?

Reply:
Looks like the published_at field was required in the model but not in the RequestDTO interface, I'll make the change in the model and update the review service accordingly.